### DISCUSSION TITLE
Modules not created with their correct names when using "Build From Source"

### DISCUSSION THREAD

**Discussion by ollypsilon** [joyfullservice/msaccess-vcs-addin#377] [Q&A] [answered]
Hi!
I want to use/test the dev branch version.
I've create a fork and a local repos with a custom brach name as described in [Editing and Contributing](https://github.com/joyfullservice/msaccess-vcs-integration/wiki/Editing-and-Contributing).
When I use the master version 3.4.23 to create the dev version like described [here](https://github.com/joyfullservice/msaccess-vcs-integration/blob/dev/CONTRIBUTING.md) 
_"Use the add-in to Build From Source, selecting the cloned Version Control.accda.src folder."_ I have lots of modules not created with their correct names, compilation fails due to this.
![image](https://user-images.githubusercontent.com/52780811/214302291-545b9527-eb72-4ff2-80ac-33f40ab386c9.png)
All failed modules have these strange characters "ï»¿" at the beginning of their exported name, e.g. _ï»¿Attribute VB_Name = "modAddInMenu"_.
Is it my mistake and I did something wrong, is it a bug in module export to .bas or what causes this?

And I found this in Build.log:
Importing modules...          
  modAddInMenu.bas
ERROR: Build error in: [...]\git repositories\msaccess-vcs-integration\Version Control.accda.src\modules\modAddInMenu.bas
Error 2489: Das modAddInMenu-Objekt ist nicht geöffnet. Source: modImportExport.Build

Is it related?

Best

Oliver

**Comment by ollypsilon** [comment]
I think it might be due to .bas files on my system are encoded as UTF-8 and not ANSI and Access seems to need ANSI encoded files during import. Because the same issue happens when I manually import the .bas files to a project.

**Comment by joyfullservice** [comment, marked as answer]
@ollypsilon - The `dev` version includes a number of enhancements and improvements over the current master branch, so I decided not to try to take the time to build all the forward-compatibility into the `master` branch. The `dev` branch builds easily from any recent `dev` version of the add-in, but obviously you need a compiled version to start with.  😄

To answer your question on the modules, after much discussion and review, the most robust source code export seemed to be that of using a UTF-8 BOM so that software applications would universally recognize the encoding of the VBA source content.

I think maybe what might be the easiest solution here is for me to publish a "pre-release" version from the `dev` branch. Then anyone wanting to test out the development version can just download it from the releases page. This pre-release version could be used to build the `dev` branch, or simply to test out some of the new features.

I just published a [binary release](https://github.com/joyfullservice/msaccess-vcs-integration/releases) of the `dev` version 4.0.5-beta. After installing this version, you can test out the features in your projects. You can also use this to build the `dev` branch. Let me know how this works for you!

**Reply by ollypsilon** [reply]
@joyfullservice Thanks for taking the time to build 4.0.5-beta. Installation was fine.
I will use this version for my project and see if it works.
But the issue with the weird characters still occur even the .bas files are encoded in UTF-8 BOM.
![image](https://user-images.githubusercontent.com/52780811/214502481-4cdd3448-3d3a-4902-8032-51c2676e3fc7.png)
This module has the error described above:
![image](https://user-images.githubusercontent.com/52780811/214503385-259f8edc-484b-43ae-8dc3-827dad4962ec.png)
But when I add some simple characters to the file the name is extracted correct and module named accordingly.
![image](https://user-images.githubusercontent.com/52780811/214504117-efd0cdf8-2167-454c-9294-f98c463d47d3.png)
![image](https://user-images.githubusercontent.com/52780811/214504150-d2d4d2b1-b09b-4af9-830d-bc368114ee54.png)
You can see it also in Build.log:
![image](https://user-images.githubusercontent.com/52780811/214504635-8fbbb861-13ae-43e4-83e3-b4e60893f498.png)
I will test it on another system to see if it's a local issue.

**Reply by joyfullservice** [reply]
Hi Oliver,

There are several things happening here, so let me explain a little about the process. 😄  In order for the source file to show non-ascii characters correctly in other programs (such as diff tools, SourceTree, etc...) the files needed to be saved as UTF-8. To provide better consistency with the source files, we decided to go ahead and use the UTF-8 BOM encoding for ALL text-based source files. If you open the source file directly in Notepad++, it should be encoded as UTF-8 BOM.

![image](https://user-images.githubusercontent.com/7254992/214611220-a1b3151b-b784-4601-90ba-b8a0c1486935.png)

If the source file in the repository is not using this encoding, that is a problem that we should investigate further.

When the file is IMPORTED into the database during the build operation, the file is converted back to the local system encoding, which may be different from the system encoding of the computer used to originally generate the source file. This is where UTF-8 is very helpful because it provides a universal standard format for the source code that can be built on machines that user other system encodings. (Such as German)

In `clsDbModule.IDbComponent_Import` you will find the following code that converts the Unicode content back to the system encoding before importing the module into VBA.

```vba
    ' Parse source file
    strName = GetObjectNameFromFileName(strFile)
    udtFile = ParseSourceFile(strFile, strName)
    
    ' Write to a new file using system encoding (converting from UTF-8)
    strTempFile = GetTempFile
    WriteFile udtFile.strContent, strTempFile, GetSystemEncoding
    
    ' Import the source code
    LoadVbeModuleFromFile strTempFile, strName
```

Note the `GetSystemEncoding` function called in the above code. This is found in `modEncoding` and attempts to map the system code page to a system encoding identifier. It would be interesting to see what your system returns if you run that function from the immediate window:

![image](https://user-images.githubusercontent.com/7254992/214614446-78f010a2-fa59-4d43-9c01-9bdb7a7815a9.png)

My hunch is that on your system the module files are not being converted back to the correct system encoding before import, which results in the BOM (byte order mark) being included as text in the file. Since the first line of the file is what specifies the module name, those extra characters cause the line to not be parsed correctly during the VBE import.

This would explain both the extra characters in the imported module, and the module name not being set when the file is imported.

The great thing about this add-in is that you can actually debug the add-in itself, adding break points and stepping through the code execution. Hopefully this will allow you to identify what is happening with the code page detection on your system. There are some notes on the `GetSystemEncoding` function that may be helpful in troubleshooting this issue.

```vba
'---------------------------------------------------------------------------------------
' Procedure : GetSystemEncoding
' Author    : Adam Waller
' Date      : 7/1/2021
' Purpose   : Return the current encoding type used for non-UTF-8 text files.
'           : (Such as VBA code modules.)
'           : https://docs.microsoft.com/en-us/windows/win32/intl/code-page-identifiers
'           : https://documentation.help/MS-Office-VB/ofhowConstants.htm
'           : * Note that using utf-8 as a default system encoding may not work
'           : correctly with some extended characters in VBA code modules. The VBA IDE
'           : does not support Unicode characters, and requires code pages to display
'           : extended/non-English characters. See Issues #60, #186, #180, #246
'---------------------------------------------------------------------------------------
'
Public Function GetSystemEncoding() As String
```

Let me know what you find out!

**Reply by ollypsilon** [reply]
Thx for the detailed process.
I hit the "beta" trap as described in module:
![image](https://user-images.githubusercontent.com/52780811/214848250-a933b515-a884-4982-923e-440649ff8f53.png)
Maybe an idea would be to create a kind of warning during export/import of UTF-8 is detected.

Now it's fine with this encoding:
![image](https://user-images.githubusercontent.com/52780811/214848034-afae7fe3-05dc-4f63-9542-8610c9b110f5.png)

Like this is was able to create the dev branch from source and also export and "build from source" my database 👍 
There's a small error on "build from source" but I'll create another discussion for it if I can't solve myself.

**Reply by joyfullservice** [reply]
Thanks for the update! Your feedback gave me an idea. Since this is really only an issue when importing VBA code modules, I have added an optional parameter in the codepage detection to fall back to a non-UTF-8 codepage when a user has the Beta Unicode option checked in Windows 10. (This fallback is only used for importing code modules, since the VBE does not have inherent UTF-8 support.) I haven't tested this because I don't use the beta option on my system, but I think this should allow the add-in to work correctly even if that option is checked.

We have had at least one other user that actually needed this beta option for a database that used Unicode characters on forms, so I think this may be a help for cases like that as well. For now I am hard-coding the `Windows-1252` fallback, but this could be extended in the future to support other code pages through a user-selected option.

If you build the `dev` branch from source, restart Access, then double-click the newly built add-in file to install it, you can test it out on your system to see if you can successfully build projects even with the beta option checked.

**Reply by ollypsilon** [reply]
So, I've built the add-in from your updated files and installed it. No issue with it.
But for my database I can't create it from source. When exporting with UTF-8 beta option checked .bas files are unreadable and cause errors during build. With UTF-8 beta option disabled export is fine, .bas file are human readable but import has issues.
Some tables are empty and I get a bunch off error messages in Build.log and in Importerrors table.
Feel free to drop my a PN if you need details or if you want me to share details here.

**Reply by joyfullservice** [reply]
@ollypsilon - Are you using actual UTF-8 characters in your VBA source code? I don't think the VBE technically supports Unicode, but I don't know exactly how it works with the UTF-8 beta option selected... Perhaps you could try exporting a module manually through VBE, then open the file in Notepad++ and check the encoding. It may be that I need to account for a different encoding there.

If you would like to create a private repository on GitHub, and add me to it, I could take a quick look at your database and source files, and see if I can identify what is happening there. On most systems the add-in seems to work right out of the box, but it is nice to solve some of those edge cases for the benefit of others that might encounter similar issues in the future.

**Reply by ollypsilon** [reply]
@joyfullservice Well, maybe a misunderstanding: it's not about VBA code exports, it's all export like e.g. table definition's .bas file which is unreadable when UTF-8 beta option is set. Without the option the .bas file is fine. On build from source the tables are generated but loading data for the ones which are exported with data fails, maybe due to UTF-8 chars in data. I will check this.
Unfortunately I'm not allowed to share the whole project due to company policy. I'm going to create a kind of debug database with just some content, allowing to reproduce the errors we got.

**Reply by joyfullservice** [reply]
Thanks, @ollypsilon! I was able to look at the sample database, and identified four issues. I will probably break these out into individual issues so they can be worked through individually.

- Content encoding issue when exporting form objects when using the UTF-8 Beta option in system language settings.
- Error parsing dates coming from different locales. (Perhaps determine a standard format that works on all systems.)
- Issue with `GetUncPath` function, resolved with 3b9809c.
- Problem importing table data when calculated fields are involved. (I have tested a solution to resolve this.) #379 
- #380 

The good news is that aside from these issues, I was able to successfully build your sample database.  😄  I am pretty sure that together we can come up with a solution that will work for the UTF-8 Beta option (which I would like to support, if we can). The other issues all seem to be things we can resolve without too much difficulty.
